### PR TITLE
In hipFree, synchronize owner of memory

### PR DESCRIPTION
**Pull request by jeffdaily**
This changes the behavior from synchronizing the currently set TLS device.

### CODE REVIEW

**Comment by gargrahul** [comment]
@jeffdaily I think we need to have an all devices level sync in hipFree. It is more critical for _appId=-1 case.

File: src/hip_memory.cpp [hunk 1]
```diff
@@ -1901,6 +1897,22 @@ hipError_t hipFree(void* ptr) {
         am_status_t status = hc::am_memtracker_getinfo(&amPointerInfo, ptr);
         if (status == AM_SUCCESS) {
             if (amPointerInfo._hostPointer == NULL) {
+                if (amPointerInfo._appId != -1) {
```
Comment (gargrahul) [inline src/hip_memory.cpp]: Do we still need this check? Since we are synchronizing all devices for now.

**Comment by gargrahul** [comment]
@jeffdaily Shall we synchronize all devices always in hipFree for now and optimize to synchronize only peer devices later separately?

File: src/hip_memory.cpp (hunk 1 above)
Comment (jeffdaily) [reply src/hip_memory.cpp]: I thought this was differentiating between the case where the pointer is owned and not shared versus when it is owned but shared.  Perhaps I misunderstood HCC pointer management and what the _appId field is used for.  I don't want hipFree() to become a performance bottleneck if we can help it.

Comment (gargrahul) [reply src/hip_memory.cpp]: _appId= -1 is used for host allocations. I don't think hipFree should be performance critical code.

Comment (gargrahul) [reply src/hip_memory.cpp]: Synchronizing all devices may help avoiding future mGPU related issues when devices are sharing pointers and freeing across devices. Does blanket all device synchronization resolve 183319?

**Comment by jeffdaily** [comment]
For now, we should synchronize the owning device.  The all-device-all-stream sync will also solve the current problem, but is too extreme of a solution at this time.  However, to help debug possible mGPU issues in the future, I have added the environment variable HIP_SYNC_FREE to enable the all-device-all-stream sync.

**Comment by bensander** [comment]
Thanks Jeff!

**Comment by gargrahul** [comment]
LGTM. We may need to revisit this when we see a real use case where HIP_SYNC_FREE helps to make all/peer devices sync as default.

**Comment by gargrahul** [comment]
@mangupta This PR needs to get merged today as this is required for ROCm 2.4 release.

**Pull request merged**